source/coap_connection_handler.c

```diff
@@ -406,27 +452,63 @@ static int timer_status(int8_t timer_id)
     return TIMER_STATE_CANCELLED;
 }
 
-static int read_data(socket_callback_t *sckt_data, internal_socket_t *sock, ns_address_t *src_address)
+static int read_data(socket_callback_t *sckt_data, internal_socket_t *sock, ns_address_t *src_address, uint8_t dst_address[static 16])
 {
     sock->data_len = 0;
     if (sckt_data->event_type == SOCKET_DATA && sckt_data->d_len > 0) {
-        if( sock->data ){
+        uint8_t ancillary_databuffer[NS_CMSG_SPACE(sizeof(ns_in6_pktinfo_t))];
+        ns_iovec_t msg_iov;
+        ns_msghdr_t msghdr;
+        ns_in6_pktinfo_t *pkt;
+
+        if (sock->data) {
             ns_dyn_mem_free(sock->data);
             sock->data = NULL;
         }
+
         sock->data = ns_dyn_mem_temporary_alloc(sckt_data->d_len);
-        if( !sock->data ){
+        if (!sock->data) {
             sock->data = NULL;
             return -1;
         }
-        sock->data_len = socket_read(sckt_data->socket_id, src_address, sock->data, sckt_data->d_len);
-    }
-    if( sock->data_len < 1){
-        ns_dyn_mem_free(sock->data);
-        sock->data = NULL;
-        sock->data_len = 0;
-        return -1;
+
+        msghdr.msg_name = src_address;
+        msghdr.msg_namelen = sizeof(ns_address_t);
+        msghdr.msg_iov = &msg_iov;
+        msghdr.msg_iovlen = 1;
+        msghdr.msg_control = ancillary_databuffer;
+        msghdr.msg_controllen = sizeof(ancillary_databuffer);
+        msghdr.flags = 0;
+
+        msg_iov.iov_base = sock->data;
+        msg_iov.iov_len = sckt_data->d_len;
+
+        sock->data_len = socket_recvmsg(sckt_data->socket_id, &msghdr, 0);
+
+        if (sock->data_len >= 0) {
+            ns_cmsghdr_t *cmsg = NS_CMSG_FIRSTHDR(&msghdr);
+
+            while (cmsg) {
+                switch (cmsg->cmsg_type) {
+                    case SOCKET_IPV6_PKTINFO:
+                        pkt = (ns_in6_pktinfo_t*)NS_CMSG_DATA(cmsg);
+                        break;
+                    default:
+                        break;
+                }
+                cmsg = NS_CMSG_NXTHDR(&msghdr, cmsg);
+            }
+            if (pkt) {
+                memcpy(dst_address, pkt->ipi6_addr, 16);
```

Comment:
You should presumably do something if !pkt - return an error, set dst_address to ns_inaddr_any? Error should be fine - this API should work.